src/init/features/turtles/repo.ts

```diff
@@ -0,0 +1,168 @@
+  if (resp.repositories.length === 0) {
+    throw new FirebaseError(
+      "The GitHub App does not have access to any repositories. Please configure" +
+        "your app installation permissions at https://github.com/settings/installations."
+    );
+  }
+
+  const choices = resp.repositories.map((repo: gcb.Repository) => ({
+    name: extractRepoSlugFromURI(repo.remoteUri) || repo.remoteUri,
+    value: repo.remoteUri,
+  }));
+  const remoteUri: string = await promptOnce({
+    type: "list",
+    message:
+      "Which of the following repositories would you like to link? If you don't" +
+      "see the repository, cancel the setup process by pressing Ctrl-C, configure" +
+      "your app installation permissions at https://github.com/settings/installations," +
+      "then run the command again.",
+    choices,
+  });
+
+  const repo = await getOrCreateRepository(projectId, location, connectionId, remoteUri);
+  logger.info(`Successfully linked GitHub repository at remote URI ${remoteUri}.`);
+  return repo;
+}
+
+// function detectGitRemote(remoteName: string): string {
+//   return execSync(`git config remote.${remoteName}.url`).toString();
+// }
+
+function extractRepoSlugFromURI(remoteUri: string): string | undefined {
+  const match = /github.com\/(.+).git/.exec(remoteUri);
+  if (!match) {
+    return undefined;
+  }
+  return match[1];
+}
+
+function generateConnectionId(): string {
+  return "turtles-conn";
+}
+
+function generateRepositoryId(remoteUri: string): string | undefined {
+  return extractRepoSlugFromURI(remoteUri)?.replace("/", "--");
+}
+
+/**
+ *
+ */
+export async function getOrCreateConnection(
+  projectId: string,
+  location: string,
+  connectionId: string
+): Promise<gcb.Connection> {
+  let conn: gcb.Connection;
+  try {
+    conn = await gcb.getConnection(projectId, location, connectionId);
+  } catch (err: unknown) {
+    if ((err as FirebaseError).status === 404) {
+      const op = await gcb.createConnection(projectId, location, connectionId);
+      conn = await poller.pollOperation<gcb.Connection>({
+        ...gcbPollerOptions,
+        pollerName: `create-${location}-${connectionId}`,
+        operationResourceName: op.name,
+      });
+    } else {
+      throw err;
+    }
+  }
+
+  // We prompt users to select Continue once they have followed
+  // the link and successfully authorized Cloud Build to access
+  // their GitHub account
+  while (conn.installationState.stage === "PENDING_USER_OAUTH") {
+    logger.info(conn.installationState.message);
+    await open(conn.installationState.actionUri);
+    const authorized = await promptOnce({
+      type: "list",
+      message:
+        "Choose 'Continue' once you have authorized Turtles (Cloud Build) to access your GitHub repo, or cancel.",
+      choices: [
+        {
+          name: "Continue",
+          value: "continue",
+        },
+        {
+          name: "Cancel",
+          value: "cancel",
+        },
+      ],
+    });
```

Comment:
Hmm I think I might consider extracting this out as a separate function.

A single function that handles API calls, polls, and user prompts smells a little.

Think it's easier for me to read code that codes like:

```
getConnection
if (stage == PENDING USER OAUTH)
  open browser
poll until state is updated (?)
```

It'd at least be eaiser to test out each components?